ForumSessionProvider/ForumSessionProvider.php

```diff
@@ -18,284 +18,851 @@
-
-        $this->userWiki->setOption('forum_last_update', time());
-        $this->userWiki->saveSettings();
-    }
+
+		// Cleanup the username.
+		$this->FSuserName = $this->cleanupUserName($this->FSuser['member_name']);
+
+		// Get all of our Forum Software groups.
+		$this->FSgroups = explode(',', $this->FSuser['additional_groups']);
+		$this->FSgroups[] = $this->FSuser['id_group'];
+
+		// Try to access this user.
+		$this->userInfo = UserInfo::newFromName($this->FSuserName, true);
+		$this->userWiki = $this->userInfo->getUser();
+
+		// If they are not logged in or the username doesnt match.
+		if (!($this->userWiki->isLoggedIn() && $this->userWiki->getName() === $this->FSuserName))
+		{
+			$this->userWiki->setId($this->userWiki->idFromName($this->FSuserName));
+
+			// The user doesn't exist yet in the wiki? Create them.
+			if ($this->userWiki->getID() === 0)
+				$this->createWikiUser();
+		}
+
+		// Make sure if we have a id match, its valid.
+		if ($this->userWiki->getOption('forum_member_id', 0) != 0 && $this->userWiki->getOption('forum_member_id', 0) != $this->FScookie['id'])
+		{
+			$this->logger->warning('Member ID ' . $this->FScookie['id'] . ' failed to match forum id. Remote IP:' . $_SERVER['REMOTE_ADDR']);
+			return null;
+		}
+
+		// Check the ban status here.
+		if ($this->isBanned())
+			return null;
+
+		// Configure all of our groups, but only every 15 minutes.
+		if (time() > ((int) $this->userWiki->getOption('forum_last_update_groups', 0) + 900))
+			$this->setWikiGroups();
+
+		/**
+		 * The forum is responsible for e-mails, so we don't need to worry about it changing much.
+		 * Updates every 30 minutes.
+		 */
+		if (time() > ((int) $this->userWiki->getOption('forum_last_update_user', 0) + 1800))
+			$this->updateWikiUser();
+
+		// Denied Login?
+		if (array_intersect($this->FSSettings['LoginDeniedGroups'], $this->FSgroups) != array())
+			return null;
+
+		// Not apart of a login group?
+		if (
+			array_intersect($this->FSSettings['LoginAllowedGroups'], $this->FSgroups) == array()
+			&& array_intersect($this->FSSettings['AdminGroups'], $this->FSgroups) == array()
+		)
+			return null;
+
+		// This was in the original code and sessionCookieName is not defined anywhere.
+		if ($this->sessionCookieName === null)
+		{
+			$id = $this->hashToSessionId($this->FSuserName);
+			$persisted = false;
+			$forceUse = true;
+		}
+		else
+		{
+			$id = $this->getSessionIdFromCookie($request);
+			$persisted = $id !== null;
+			$forceUse = false;
+		}
+
+		// Stand up a new session for MediaWiki.
+		return new SessionInfo(SessionInfo::MAX_PRIORITY, array(
+			'provider' => $this,
+			'id' => $id,
+			'userInfo' => $this->userInfo,
+			'persisted' => $persisted,
+			'forceUse' => $forceUse,
+		));
+	}
+
+	/*
+	 * Fetch the database information for a user.
+	*/
+	private function getFSUser(WebRequest $request)
+	{
+		global $wgRequest;
+
+		// !!! TODO: Caching this?  Can't use $_SESSION ANYMORE.

```

Comment:
I was looking into that, but couldn't find any good simple examples of the caching.  I find MediaWiki code hard to read as its overly complex.  I couldn't find a straight forward example of the caching process.

But it would be ideal to cache this.  Every time the user hits the page, it decodes the cookie, connects to the FS database and issues a query against it.  Seems to me if we can cache that, we could issue fewer queries.  Especially if we can cache it in a users session.  So while they are browsing the wiki, it maybe only hits the FS database 1-2 times in a 15 minute window.

Reply:
Not disagreeing with the need to cache, though it's a bit of a lesser concern for me. Should probably straight make an issue for caching this and bring discussion over.